pkg/logging/rate_limited_logger.go

```diff
@@ -0,0 +1,73 @@
+package logging
+
+import (
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"golang.org/x/time/rate"
+)
+
+// RateLimitedLogger wraps a zap.SugaredLogger that supports rate limiting.
+type RateLimitedLogger struct {
+	*Logger
+	rateLimiters *lru.Cache
+	// needs a mutex to protect burst count increase? //@TODO: khushboo
+}
+
+// NewRateLimitLogger returns a rate limited logger
+func NewRateLimitLogger(l *Logger, size int) *RateLimitedLogger {
+	cache, err := lru.New(size)
+	if err != nil {
+		l.Errorf("unable to create rate limiter cache for logger in module %q: %q", l.module.name, err)
+		return nil
+	}
+	return &RateLimitedLogger{
+		l,
+		cache,
+	}
+}
+
+// SetLimiter sets how many logLines are emitted in a given interval for a specific identifier.
+func (rl *RateLimitedLogger) SetLimiter(limiter string, logLines float64, interval time.Duration, burst int) {
+	if logLines == float64(burst) {
+		burst++
+	}
+	// Check first, and only add the limiter if it doesn't exist already.
+	_, _ = rl.rateLimiters.ContainsOrAdd(limiter, rate.NewLimiter(rate.Limit(logLines/interval.Seconds()), burst))
+}
+
+// ErrorL logs a templated error message if allowed by the rate limiter corresponding to the identifier
+func (rl *RateLimitedLogger) ErrorL(limiter string, template string, args ...interface{}) {
+	if rl.allowLog(limiter) {
+		rl.Errorf(template, args...)
+	}
```

Comment:
I'd add an `else` path here and below printing a—also rate-limited!—note saying that log entries are being rate-limited.

Reply:
Wouldn't that defeat the purpose of rate limiting the log messages in the first place?

Reply:
So generally the rate limiter system would log when things are being rate limited. Otherwise it's hard to tell if it's due to rate limits vs low traffic vs bug. 

But the challenge here is that we are rate limiting logs, so logging that is counter-productive.

I can think of a few alternatives:

1. Log that we are rate limiting in _debug_ mode. Pros: No log spam in prod, we can still see logs when debugging. Con: If we run into an issue in prod, we can't tell without switching logs to debug mode. Debug logs are noisy.

2. Create a new (global) rate limiter, and use that to log something like "Rate limiting error logs including this log". 

3. Write the "Rate limiting error logs" message on a timer every 5 minutes.

I'm leaning towards #1 purely because it's easy to comprehend. Thoughts?

Reply:
+1, on 1. Other options are overkill in the light of an ingress rate limiter which has to be implemented for GA.